src/db.c

```diff
@@ -1604,25 +1604,143 @@ int *getKeysPrepareResult(getKeysResult *result, int numkeys) {
     if (numkeys > result->size) {
         if (result->keys != result->keysbuf) {
             /* We're not using a static buffer, just (re)alloc */
-            result->keys = zrealloc(result->keys, numkeys * sizeof(int));
+            result->keys = zrealloc(result->keys, numkeys * sizeof(keyReference));
         } else {
             /* We are using a static buffer, copy its contents */
-            result->keys = zmalloc(numkeys * sizeof(int));
+            result->keys = zmalloc(numkeys * sizeof(keyReference));
             if (result->numkeys)
-                memcpy(result->keys, result->keysbuf, result->numkeys * sizeof(int));
+                memcpy(result->keys, result->keysbuf, result->numkeys * sizeof(keyReference));
         }
         result->size = numkeys;
     }
 
     return result->keys;
 }
 
+/* Fetch the keys based of the provided key specs. Returns 0 if the keyspec definition is
+ * incomplete. */
+int getKeysUsingKeySpecs(struct redisCommand *cmd, robj **argv, int argc, getKeysResult *result) {
+    int j, i, k = 0, last, first, step;
+    keyReference *keys;
+
+    for (j = 0; j < cmd->key_specs_num; j++) {
+        keySpec *spec = cmd->key_specs+j;
+        serverAssert(spec->begin_search_type!=KSPEC_BS_INVALID);
+        if (spec->flags & CMD_KEY_INCOMPLETE) {
+            result->numkeys = 0;
+            return 0;
+        }
+
+        first = 0;
+        if (spec->begin_search_type == KSPEC_BS_INDEX) {
+            first = spec->bs.index.pos;
+        } else { /* KSPEC_BS_IKEYWORD */
+            int start_index = spec->bs.keyword.startfrom > 0 ? spec->bs.keyword.startfrom : argc+spec->bs.keyword.startfrom;
+            int end_index = spec->bs.keyword.startfrom > 0 ? argc-1: 0;
```

Comment:
```suggestion
            int end_index = spec->bs.keyword.startfrom > 0 ? argc-1: 1;
```
i suppose we don't need to look for the keyword in the command name.
and as already coded, we don't look for it in the very last arg (since the keys are expected to follow)..